client/post-editor/editor-publish-button/index.jsx

```diff
@@ -0,0 +1,123 @@
+/**
+ * External dependencies
+ */
+import React, { PropTypes } from 'react';
+import PureRenderMixin from 'react-pure-render/mixin';
+
+/**
+ * Internal dependencies
+ */
+import localize from 'lib/mixins/i18n/localize';
+import stats from 'lib/posts/stats';
+import postUtils from 'lib/posts/utils';
+import siteUtils from 'lib/site/utils';
+
+export const EditorPublishButton = React.createClass( {
+	propTypes: {
+		site: PropTypes.object,
+		post: PropTypes.object,
+		savedPost: PropTypes.object,
+		onSave: PropTypes.function,
+		onPublish: PropTypes.function,
+		tabIndex: PropTypes.number,
+		isPublishing: PropTypes.bool,
+		isSaveBlocked: PropTypes.bool,
+		hasContent: PropTypes.bool
+	},
+
+	mixins: [ PureRenderMixin ],
+
+	trackClick: function() {
+		const postEvents = {
+			update: 'Clicked Update Post Button',
+			schedule: 'Clicked Schedule Post Button',
+			requestReview: 'Clicked Request-Review Post Button',
+			publish: 'Clicked Publish Post Button',
+		};
+		const pageEvents = {
+			update: 'Clicked Update Page Button',
+			schedule: 'Clicked Schedule Page Button',
+			requestReview: 'Clicked Request-Review Page Button',
+			publish: 'Clicked Publish Page Button',
+		};
+		const buttonState = this.getButtonState();
+		const eventString = postUtils.isPage( this.props.post ) ? pageEvents[ buttonState ] : postEvents[ buttonState ];
+		stats.recordEvent( eventString );
+		stats.recordEvent( 'Clicked Primary Button' )
+	},
+
+	getButtonState: function() {
+		if (
+			postUtils.isPublished( this.props.savedPost ) &&
+			! postUtils.isFutureDated( this.props.post ) ||
+			(
+				this.props.savedPost &&
+				this.props.savedPost.status === 'future' &&
+				postUtils.isFutureDated( this.props.post )
+			)
+		) {
+			return 'update';
+		}
+
+		if ( postUtils.isFutureDated( this.props.post ) ) {
+			return 'schedule';
+		}
+
+		if ( siteUtils.userCan( 'publish_posts', this.props.site ) ) {
+			return 'publish';
+		}
+
+		if ( this.props.savedPost && this.props.savedPost.status === 'pending' ) {
+			return 'update';
+		}
+
+		return 'requestReview';
+	},
+
+	getButtonLabel: function() {
+		const buttonLabels = {
+			update: this.props.translate( 'Update' ),
```

Comment:
Probably a micro-optimization, but translate is a non-trivially expensive operation, so I wonder if we could rewrite the function to invoke it only once per call. `switch` (or a set of `if` conditions) would be a good choice:

```js
getButtonLabel: function() {
	switch ( this.getButtonState() ) {
		case 'update': return this.props.translate( 'Update' );
		case 'schedule': return this.props.translate( 'Schedule' );
		case 'publish': return this.props.translate( 'Publish' );
		case 'requestReview': return this.props.translate( 'Submit for Review' );
	}
}
```